tunneler/tunneler.py

```diff
@@ -0,0 +1,98 @@
+from socket import *
+from tunneler.network_headers import EtherHeader, IpHeader
+import threading
+
+L3_PROTO_IP = 0x0800
+MAX_BUF_SIZE = 0xffffffff
+
+
+def _create_raw_ip_socket(interface):
+    """
+    Create a raw socket to listen for ip packets on a specific interface
+    :param interface: the interface for the created socket to listen on
+    :return: a socket object
+    NOTE: the socket will receive the layer 2 and 3 headers, in addition to the payload
+    """
+    raw_sock = socket(AF_PACKET, SOCK_RAW, L3_PROTO_IP)
+    #raw_sock.setsockopt(SOL_SOCKET, SO_RCVBUF, MAX_BUF_SIZE)
+
+    raw_sock.bind((interface, L3_PROTO_IP))
+
+    return raw_sock
+
+class L2Tunnel:
+    """
+    A class for handling tunnel forwarding on layer 2
+    """
+    def __init__(self, target_mac, gateway_mac, my_mac, interface):
+        self.target_mac = target_mac
+        self.gateway_mac = gateway_mac
+        self.my_mac = my_mac
+        self.raw_sock = _create_raw_ip_socket(interface)
+        self.should_forward = False
+        self.forward_thread = threading.Thread(target=self.forward_loop, args=())
```

Comment:
two things:
* should be private
* if you do not create thread object every time start tunnel is called, you will only be able to start and stop once